core/shared/src/main/scala/fs2/Pull.scala

```diff
@@ -70,83 +210,147 @@ sealed abstract class Pull[+F[_], +O, +R] {
-          case res @ Interrupted(_, _) => res
-          case res @ Fail(_)           => res
+          case other => other
         }
     }
 
-  /** Alias for `flatMap(_ => p2)`. */
-  def >>[F2[x] >: F[x], O2 >: O, R2](p2: => Pull[F2, O2, R2]): Pull[F2, O2, R2] =
-    new Bind[F2, O2, R, R2](this) {
-      def cont(r: Terminal[R]): Pull[F2, O2, R2] =
-        r match {
-          case _: Succeeded[_] => p2
-          case r: Interrupted  => r
-          case r: Fail         => r
-        }
-    }
+  /** Run `post` after `this`, regardless of errors during `this`: 
+    *
+    * - If `this` pull completes successfully, the return value is discarded
+    *   and the `post` pull is evaluated. The return value of the `post` pull 
+    *   be it success, error, or interruption, is the end result of the combined pull.
+    *
+    * - If `this` pull is aborted with a failure, then the  `post` pull is 
+    *   compiled next. If the `post` pull ends up with an error or interruption,
+    *   that will be the result of the combined pull. However, if the `post` pull 
+    *   completes successfully, then the combined pull rethows the `err` error as its result.
+    *
+    * - If the `this` pull is interrupted, then the `post` error is never run
+    *   and the combined pull return with that interruption.
+    */
+  def onComplete[G[x] >: F[x], O2 >: O, R2](post: => Pull[G, O2, R2]): Pull[G, O2, R2] =
+    handleErrorWith(e => post >> Fail(e)) >> post
 
-  /** Lifts this pull to the specified effect type. */
-  def covary[F2[x] >: F[x]]: Pull[F2, O, R] = this
+  /** Short-hand for `(this: Pull[G, P, S])`
+    * useful to override compiler type inference.
+    */
+  def covaryAll[G[x] >: F[x], O2 >: O, R2 >: R]: Pull[G, O2, R2] = this
 
-  /** Lifts this pull to the specified effect type, output type, and resource type. */
-  def covaryAll[F2[x] >: F[x], O2 >: O, R2 >: R]: Pull[F2, O2, R2] = this
+  /** Short-hand for `(this: Pull[G, O, R])`,
+    * useful to override compiler type inference.
+    */
+  def covary[G[x] >: F[x]]: Pull[G, O, R] = this
 
-  /** Lifts this pull to the specified output type. */
+  /** Short-hand for `(this: Pull[F, O2, R])`
+    * useful to override compiler type inference.
+    */
   def covaryOutput[O2 >: O]: Pull[F, O2, R] = this
 
-  /** Lifts this pull to the specified resource type. */
-  def covaryResource[R2 >: R]: Pull[F, O, R2] = this
+  /** Short-hand for `(this: Pull[F, O, R2])`
+    * useful to override compiler type inference.
+    */
+  def covaryResult[R2 >: R]: Pull[F, O, R2] = this
+
+  /** Returns a pull with the result wrapped in `Right`, or an error wrapped in `Left` if the pull has failed. */
+  def attempt: Pull[F, O, Either[Throwable, R]] =
+    map(r => Right(r)).handleErrorWith(t => Succeeded(Left(t)))
 
-  /** Applies the resource of this pull to `f` and returns the result in a new `Pull`. */
+  /** Transforms the carry-over result type of this pull to `f` and returns the result in a new `Pull`.
```

Comment:
Like above, avoid mention of compilation. Also, avoid "carry-over" terminology.

Reword to something like:

```
Converts the result of this pull using the supplied function.
```